ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/pull-queries-against-materialized-table-table-joins.json

```diff
@@ -0,0 +1,466 @@
+        "SELECT * FROM INNER_JOIN;"
+      ],
+      "inputs": [
+        {"topic": "left_topic", "key": 0, "value": {"NAME": "zero", "VALUE": 0}, "timestamp": 0},
+        {"topic": "right_topic", "key": 0, "value": {"F1": "blah", "F2": 50}, "timestamp": 10000},
+        {"topic": "left_topic", "key": 10, "value": {"NAME": "100", "VALUE": 5}, "timestamp": 11000},
+        {"topic": "left_topic", "key": 0, "value": {"NAME": "foo", "VALUE": 100}, "timestamp": 13000},
+        {"topic": "right_topic", "key": 0, "value": {"F1": "a", "F2": 10}, "timestamp": 15000},
+        {"topic": "right_topic", "key": 15, "value": {"F1": "c", "F2": 20}, "timestamp": 15500},
+        {"topic": "left_topic", "key": 0, "value": {"NAME": "bar", "VALUE": 99}, "timestamp": 16000},
+        {"topic": "left_topic", "key": 90, "value": {"NAME": "ninety", "VALUE": 90}, "timestamp": 17000}
+      ],
+      "responses": [
+        {"admin": {"@type": "currentStatus"}},
+        {"admin": {"@type": "currentStatus"}},
+        {"admin": {"@type": "currentStatus"}},
+        {"query": [
+          {"header":{"schema":"`T_ID` BIGINT KEY, `NAME` STRING, `VALUE` BIGINT, `F1` STRING, `F2` BIGINT"}},
+          {"row":{"columns":[0,"bar",99,"a",10]}}
+        ]}
+      ]
+    },
+    {
+      "name": "table table outer join",
+      "format": ["JSON"],
+      "statements": [
+        "CREATE TABLE TEST (ID BIGINT PRIMARY KEY, NAME varchar, VALUE bigint) WITH (kafka_topic='left_topic', value_format='{FORMAT}');",
+        "CREATE TABLE TEST_TABLE (ID BIGINT PRIMARY KEY, F1 varchar, F2 bigint) WITH (kafka_topic='right_topic', value_format='{FORMAT}');",
+        "CREATE TABLE OUTER_JOIN as SELECT ROWKEY AS ID, name, value, f1, f2 FROM test t FULL OUTER join TEST_TABLE tt on t.id = tt.id;",
+        "SELECT * FROM OUTER_JOIN;"
+      ],
+      "inputs": [
+        {"topic": "left_topic", "key": 0, "value": {"NAME": "zero", "VALUE": 0}, "timestamp": 0},
+        {"topic": "right_topic", "key": 0, "value": {"F1": "blah", "F2": 50}, "timestamp": 10000},
+        {"topic": "left_topic", "key": 10, "value": {"NAME": "100", "VALUE": 5}, "timestamp": 11000},
+        {"topic": "left_topic", "key": 0, "value": {"NAME": "foo", "VALUE": 100}, "timestamp": 13000},
+        {"topic": "right_topic", "key": 0, "value": {"F1": "a", "F2": 10}, "timestamp": 15000},
+        {"topic": "right_topic", "key": 15, "value": {"F1": "c", "F2": 20}, "timestamp": 15500},
+        {"topic": "left_topic", "key": 0, "value": {"NAME": "bar", "VALUE": 99}, "timestamp": 16000}
+      ],
+      "responses": [
+        {"admin": {"@type": "currentStatus"}},
+        {"admin": {"@type": "currentStatus"}},
+        {"admin": {"@type": "currentStatus"}},
+        {"query": [
+          {"header":{"schema":"`ID` BIGINT KEY, `NAME` STRING, `VALUE` BIGINT, `F1` STRING, `F2` BIGINT"}},
+          {"row":{"columns":[15, null, null, "c", 20]}},
```

Comment:
Hmm.. why the keys are not sorted here? Does ordering actually not matter in the test format?

Reply:
The ordering does matter in the test format. For a single partition case the order is always certain (That's why the test cases pass every time). The definitive order is determined by the state store itself. I don't know the underlying data structure that stores all the records, so I can't speak to the logic behind the ordering. It is 'random' in a definite way. 
For multiple partitions, there is no determined ordering because of async processing

Reply:
In that case, why it is ordered as 15 -> 0 -> 10? Shouldn't it be 0 -> 10 -> 15?

Reply:
The ordering is definitive, not sorted

Reply:
From my understanding on the table-table join logic, the emitted records' keys should be 0, 10, 15; hence I'm curious to learn why it is 15,0,10.

Reply:
yes, but it gets stored in a state store from which pull queries fetch the records. We don't offer any guarantees on the ordering in which the state stores serves the records to the pull query afaik. We can dig into it, but there is a plausible explanation for why the pull query ordering doesnt match the persistent query ordering.

Reply:
I guess a higher order question/task would be: do we want to offer any ordering guarantees around pull query records? Should they match persistent query ordering in some way?

Reply:
@AlanConfluent thoughts?

Reply:
Since this test is on a single partition, the ordering is dependent on that single state store's range query -- although we do not commit to any ordering today, our current implementation on the range query on persistent rocksDB does indeed return records by key order. That's why I'm curious why only this single test returns results not in key order, while all other tests seem to return results in key order.

Reply:
We generally don't have ordering guarantees since we do requests in parallel and they stream back results.  In this case, if it's a single partition, it should presumably be the ordering of the underlying store.  I'm honestly not sure why it wouldn't be sorted, though in tests, they're certainly consistent.  We could spend a few minutes looking into this if we're curious, but it doesn't seem like a big deal since we can't make guarantees anyway.

Reply:
I guess I'm just curious since 1) this result is consistent, and 2) all other test results are ordered indeed, and I cannot think of a reason why it is different from other results by just looking at the statements. But I agree that at the moment we do not provide any guarantees anyways. So this question is mainly for our own understanding.

Reply:
@guozhangwang there are other tests that have unsorted ordering as well

Reply:
I spent some more time thinking about it and still cannot come up with the potential root cause (I looked at the source code at KS too but did not find anything there). Filed https://github.com/confluentinc/ksql/issues/7391.

Reply:
Just to clarify, this should not be a blocker for merging this PR as is, we could just investigate it separately as we continue making progress on this one.